### DISCUSSION TITLE
CTF 4: CodeQL and Chill- Step 1.1

### DISCUSSION THREAD

**Discussion by Zast996** [github/securitylab#132] [General] [answered]
Hi everyone, I am trying to resolve the step 1.1 of the challenge, reported below:

**The sources of tainted data are the bean properties that go through constraint validation. In the code, these can be found as the first parameter of ConstraintValidator.isValid(...).
Write a CodeQL predicate that identifies these call arguments:
predicate isSource(DataFlow::Node source) { /* TODO describe source */ }
To test your predicate, use the Quick Evaluation command (Right-click > CodeQL: Quick Evaluation). You should get 6 results.** 

I get the following solution:
![code](https://user-images.githubusercontent.com/62074263/85125262-d1625e00-b21a-11ea-90e9-87f2d2f2797e.PNG)

but the result isn't the expected:
![solution](https://user-images.githubusercontent.com/62074263/85125322-e8a14b80-b21a-11ea-90d6-c4f681aa6797.PNG)

I tried different solutions, but all of them provide the same result.
May someone please tell me where is the problem with my solution?

**Comment by intrigus-lgtm** [comment, marked as answer]
Read carefully again:
> The sources of tainted data are the bean properties that go through constraint validation. In the code, __these can be found as the first parameter__ of ConstraintValidator.isValid(...).

If my hint is not helpful, I can give further hints or completely explain the problem in your solution.

**Reply by Zast996** [reply]
The first parameter should be get (or not?) through the following statement in the where clause:
` source.asExpr() = call.getArgument(0) `

**Reply by intrigus-lgtm** [reply]
There is a big difference between a parameter and an argument.
`source.asParameter() = ...`

**Reply by Zast996** [reply]
Thanks for the hint.
I tried with:
`source.asParameter() = call.getMethod().getParameter(0) `
The result is different but I only got 2 results when I was supposed to get 6.
![Capture](https://user-images.githubusercontent.com/62074263/85175035-29728200-b266-11ea-9f0c-278cddcff9b5.PNG)
In addition, i don't know why the parameter "Object value" isn't get as a result, although it is the first parameter of a "isValid" method inside a class which implements "ConstraintValidator<>"

**Reply by intrigus-lgtm** [reply]
For the future, please post your current query as text :)
I suggest using **Quick Evaluation** on both parts of the `and` and also on the whole.
````codeql
method.getName() = "isValid" and
method.getDeclaringType().getAnANcestor().getASupertype().getName() = " ConstraintValidator<>"
````
You're trying to find the overrides of a method, but I think that this will not work this way.
There is a `overrides` predicate that you can use.

After some time, you will probably come to the conclusion that `overrides` does not work for our use-case.
This is due to generics.
if so, have a look at https://github.com/github/codeql/issues/3490.

And as always, the greatest tool in debugging is **Quick Evaluation**.

**Reply by Zast996** [reply]
I did some tests using Quick Evaluation tool as you suggested.
I figured out that the problem was due to the method access call, so I tried with the following query:
```codeQL
predicate isSource(DataFlow::Node source) {
  exists( Method method |
    source.asParameter() = method.getParameter(0) and
    method.getName() = "isValid" and
    method.getDeclaringType().getAnAncestor().getASupertype().getName()
    = "ConstraintValidator<>" and
    )
}
```
This time the result seems almost the right one.
I got 8 results because there are 2 isValid method not overridden too, so I had a look to the link you shared.
I tried with a new query:
```codeQL
predicate isSource(DataFlow::Node source) {
  exists( Method method, Method overridden |
    source.asParameter() = method.getParameter(0) and
    method.getName() = "isValid" and
    method.getDeclaringType().getAnAncestor().getASupertype().getName()
    = "ConstraintValidator<>" and
    overridden.getSourceDeclaration() = method and
    method.overrides*(overridden) 
    )
}
```
However, the result isn't change (the same 8 results as before), but i think that the use of the overrides* method should be right.

**Reply by intrigus-lgtm** [reply]
One big problem is this:
````codeql
method.getName() = "isValid" and 
method.getDeclaringType().getAnAncestor().getASupertype().getName()
= "ConstraintValidator<>"
````
This will select all method that have the name `isValid`.
What you are trying to do here is selecting only methods that are derived from `ConstraintValidator`.
But consider these (pseudo-code-ish) classes:
````java
class Foo implements Bar {
public void isValid(String clearly, String not, String a, String constraintValidatorMethod);
}

class Bar implements ConstraintValidator {}
````
Your query would flag the above `isValid` method because it is declared in `Foo`. And `Bar` is an ancestor of `Foo`, but also `ConstraintValidator` is an ancestor of `Foo`.

Instead I'd suggest using something like this:
````codeql
isValidMethod.hasName("isValid") and
  isValidMethod.getDeclaringType().getName() = "ConstraintValidator"
````
this will get you all `isValid` method that are defined in `ConstraintValidator`.
This will give you two results. I don't know why there are two results, but it is probably related to Generics...
Then I'd use `m.overridenOrInstantiates(isValidMethod)` to find all methods `m` that override the `isValid` method from `ConstraintValidator`.

**Reply by Zast996** [reply]
Finally I managed to resolve the Step.
I used your suggestion, but keeping the getAncestor method.
```codeQL
predicate isSource(DataFlow::Node source) {
  exists( Method method, Method m |
    source.asParameter() = m.getParameter(0) and
    method.getName() = "isValid" and
    method.getDeclaringType().getAnAncestor().getASupertype().getName()
    = "ConstraintValidator<>" and
    m.overridesOrInstantiates(method)
    )
}
```
I used the method m in the getParameter function instead of isValidMethod and used your override instruction.
The query gets 6 isValidMethod overridden.
Thanks a lot :)
